### PR TITLE
Implement Envelope Q-Learning Algorithm

### PR DESCRIPTION
**Pull request by jonpsy**
Envelope Q-Learning (EQL) is a multi-objective reinforcement learning algorithm that focuses on optimizing a set of objectives rather than a single objective. The algorithm learns a deterministic single policy model, however, it can adapt to user-specified preference at the execution phase thus covering the entire preference space. 

**TODO:**

- [x] Implement ```BestAction```.
- [x] Write logic for setting ```target``` value.
- [x] Write logic for feeding input to network.
- [ ]  Complete ```PrioritizedEQLReplay```.
- [x]  Complete ```BackwardEQL```
- [ ]  Distribution of ```D_w``` should be provided by user.
- [x] Implement homotopy loss.

@zoq  @coatless @say4n @shrit 

Original paper: https://arxiv.org/abs/1908.08342.

### CODE REVIEW

**Comment by jonpsy** [comment]
The algorithm is currently in pre-alpha stage. The code has been ported from ```q_learning.hpp``` as much of it will be shared in this code as well. Stay tuned for more.

File: src/mlpack/methods/reinforcement_learning/eql.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ */
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType = PrioritizedReplay<EnvironmentType>
+>
+class EQL
+{
```
Comment (jonpsy) [inline src/mlpack/methods/reinforcement_learning/eql.hpp:59]: I want to set the distribution of the preference vector, how do I do it? For ex: whether preference weights are sampled from gaussian or gamma distribution during training.  The D_w part in the paper

image

Comment (say4n) [reply src/mlpack/methods/reinforcement_learning/eql.hpp:59]: In the *policy adaptation* part just after the algorithm definition, they mention using a multivariate gaussian with m means and a single standard deviation to sample the preference vector from. However, I am not quite sure if they use the same distribution for learning too.

IMG_7F7ED162ED5D-1

Comment (jonpsy) [reply src/mlpack/methods/reinforcement_learning/eql.hpp:59]: Yes true! The code agrees with you as well. Here's what they've done in code:

```python
        if preference is None:
            if self.w_kept is None:
                self.w_kept = torch.randn(self.model_.reward_size)
                self.w_kept = (torch.abs(self.w_kept) / \
                               torch.norm(self.w_kept, p=1)).type(FloatTensor)
            preference = self.w_kept
```

I guess this is the "assumption part", because preference distribution was ```None```.  We should equip users with the ability to set preference distribution.

Comment (say4n) [reply src/mlpack/methods/reinforcement_learning/eql.hpp:59]: We could keep it as an optional argument and default to the multi-variate gaussian too; as the authors did. :)

**Review by zoq — COMMENTED** [review]
In the gdoc you mentioned you might want to put the code into a separate folder, do you mean `src/mlpack/methods/reinforcement_learning/moq`? And put `eql.hpp` in there as well or only the utility functions?

File: src/mlpack/methods/reinforcement_learning/eql_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  size_t actionSize = EnvironmentType::Action::size;
+  size_t batchSize = actionValues.n_cols / (numWeights * actionSize);
+  size_t extendedSize = numWeights * batchSize;
+
+  // Each preference vector is repeated batchSize * actionSize
+  // number of times. Shape: (rewardSize, extendedSize * actionSize).
+  const arma::mat extWeights = [&]()
+  {
+    arma::mat retval(rewardSize, extendedSize * actionSize);
+    size_t colIdx {}, start {};
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/eql_impl.hpp:119]: We usually use `= 0` for initialization, because this will just use the default constructor and  it's not immediately clear that that means zero initialization.

File: src/mlpack/methods/reinforcement_learning/eql_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    size_t gap = batchSize * actionSize;
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/eql_impl.hpp:120]:
```suggestion
    const size_t gap = batchSize * actionSize;
```

Just to point out this is something that will not change in the while loop.

File: src/mlpack/methods/reinforcement_learning/eql_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+      start += gap;
+      ++colIdx;
+    }
+
+    return retval;
+  }();
+
+  // Batch multiply extWeights and actionValues and take the max index.
+  return arma::index_max(arma::reshape(arma::sum(extWeights % actionValues, 0),
+                                       actionSize, extendedSize), 0);
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/eql_impl.hpp:136]:
```suggestion
  return arma::index_max(arma::reshape(arma::sum(extWeights % actionValues, 0),
      actionSize, extendedSize), 0);
```

Should be tabbed twice.

File: src/mlpack/methods/reinforcement_learning/replay/prioritized_eql_replay.hpp [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    sampledIndices = SampleProportional();
+    BetaAnneal();
+
+    size_t numWeights = weightSpace.n_cols;
+    size_t batchSize = sampledIndices.n_cols;
+
+    const arma::mat batchWeights = [&]()
+    {
+      arma::mat retval(rewardSize, extendedSize);
+      size_t colIdx {}, start {};
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/replay/prioritized_eql_replay.hpp:255]: See comments above.

**Review by zoq — COMMENTED** [review]
First pass.

File: src/mlpack/methods/reinforcement_learning/eql_impl.hpp (hunk 4 above)
Comment (jonpsy) [reply src/mlpack/methods/reinforcement_learning/eql_impl.hpp:136]: Did this so that it's clear that those args are for ```arma::reshape```, the meaning would be lost if use the normal tabbings.

**Comment by jonpsy** [comment]
Won't be able to continue this for a while. If anybody likes to take this over or needs help reviewing happy to help.

File: src/mlpack/methods/ann/moqn/moqn_impl..hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    const PredictorsType& inputs,
+    const TargetsType& targets,
+    const WeightsType& extendedWeightSpace,
+    LossFunctionType HomotopyLoss,
+    GradientsType& gradients)
+{
+  double loss = 0.0;
+  std::tie(error, loss) = HomotopyLoss(
+      boost::apply_visitor(outputParameterVisitor, network.back()), target);
+
```
Comment (jonpsy) [inline src/mlpack/methods/ann/moqn/moqn_impl..hpp:38]: See how this plays out? Now it acts just like a `outputLossLayer`

File: src/mlpack/methods/ann/moqn/moqn_impl..hpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  gradients = arma::zeros<arma::mat>(parameter.n_rows, parameter.n_cols);
```
Comment (jonpsy) [inline src/mlpack/methods/ann/moqn/moqn_impl..hpp:39]: I'm a little worried about the gradient, will it work as expected assuming `error` is correct? @zoq

File: src/mlpack/methods/reinforcement_learning/eql_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    targetNetwork.Parameters() = learningNetwork.Parameters();
+
+  if (totalSteps > config.ExplorationSteps())
+    policy.Anneal();
+
+  if (totalSteps > config.LambdaUpdateSteps())
+  {
+    // Lambda anneal.
+    lambda += lambda_delta;
+    lambda_delta = (lambda - lambdaInit) * lambdaExpBase + lambdaInit - lambda;
```
Comment (jonpsy) [inline src/mlpack/methods/reinforcement_learning/eql_impl.hpp:271]: @zoq  Perhaps we could wrap it inside a policy? Like `lambdaUpdatePolicy`, it's a little awkward to hold state of `lambda`, `lambdaInit` & `lambda_delta` inside our `EQL` class.

Comment (zoq) [reply src/mlpack/methods/reinforcement_learning/eql_impl.hpp]: Either is fine, right now we only have a single update strategy, but I can see that a policy could make sense here anyway.

File: src/mlpack/methods/ann/moqn/moqn_impl..hpp (hunk 7 above)
Comment (zoq) [reply src/mlpack/methods/ann/moqn/moqn_impl..hpp:53]: I think so.
